### ISSUE TITLE
[Feature Request] Support parsing ISO 8601 Duration strings

### ISSUE THREAD

**Issue by praveentiru** [chronotope/chrono#579]
The crate supports parsing of date strings but, there is no equivalent support for parsing duration strings. This is request to support parsing of duration strings. Details here -> https://en.wikipedia.org/wiki/ISO_8601#Durations

I am open to work on this issue but, would need direction/guidance to do the same.

**Comment by praveentiru** [comment]
I was poking around the code base. I can do parsing using Parsed using similar approach as implementation of `FromStr` trait for `DateTime`. But, I am not sure what would be the right way to expose this functionality.

Look forward to feedback.

**Comment by meh** [comment]
Is anyone working on this? I kind of need it.

**Comment by esheppa** [comment]
I like the struct you suggest, but this part from wiki concerns me:
> The smallest value used may also have a decimal fraction,[[37]](https://en.wikipedia.org/wiki/ISO_8601#cite_note-37) as in "P0.5Y" to indicate half a year.

A pragmatic option may be that we only support integer values everywhere but in the seconds part, however it would be interesting to understand whether that is overly restrictive.

**Comment by djc** [comment]
Oh, good catch. I can't think of any strategy for figuring out how restrictive that would be. Starting with integer values only seems like a decent approach to get at least some support in.

**Comment by xfix** [comment]
ISO 8601 has a lot of features that aren't particularly useful. Personally I rather would look into supporting [RFC 3339](https://www.rfc-editor.org/rfc/rfc3339), which is a practical subset of it.

**Comment by djc** [comment]
Huh, Wikipedia has this:

> IETF RFC 3339[[45]](https://en.wikipedia.org/wiki/ISO_8601#cite_note-45) defines a profile of ISO 8601 for use in [Internet protocols](https://en.wikipedia.org/wiki/Internet_protocol_suite) and [standards](https://en.wikipedia.org/wiki/Internet_standard). It explicitly excludes durations and dates before the [common era](https://en.wikipedia.org/wiki/Common_era).

So it looks like RFC 3339 does in fact include durations? I'm happy to bill our support as only for the RFC 3339 subset of durations. It does look like I missed a `weeks` field in my proposed definition above.

**Comment by xfix** [comment]
Oh wait, it doesn't provide durations, I just got confused by CTRL+F in RFC 3339 providing a result in "ISO 8601 Collected ABNF" section.

**Comment by djc** [comment]
However, it seems hopeful that the informational RFC 3339 grammar for duration doesn't allow decimals at all.

**Comment by esheppa** [comment]
We could call the support "RFC 3339 inspired" :D

**Comment by Ownezx** [comment]
Yeah some support for something like that would be great. I'm currently trying to use chrono for ics file parsing and being able to parse the duration in these files (the duration have a format like so `P15DT5H0M20S`) would be great.

Having something like the `DateTime::parse_from_str` for duration would be golden.

**Comment by djc** [comment]
@Ownezx would you be interested in working on a PR along the lines of the discussion above?

**Comment by Ownezx** [comment]
I'd have to look at the source for `DateTime::parse_from_str`. But I'm rather new to rust and to contributing to open source project.

Can't give any garantee but I can give it a try for sure.

**Comment by Ownezx** [comment]
Just to make things clear, you'd want me to play with the `TimeDelta` struct or with the `time::duration` struct?

**Comment by djc** [comment]
It would be a new type, somewhat like the `CalendarPeriod` type I defined in an earlier comment.

**Comment by michaelmior** [comment]
Note that JSON Schema explicitly uses ISO 8601 format durations. Since `arrow-json` already uses `chrono`, adding support for parsing ISO 8601 durations would be an easy way to add support for durations to Arrow's JSON reader.
